eloquentjs/4_3_list.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToList(inputArray) { 
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let currentIndex = inputArray.length-1;
+
+    function innerArrayToList(inputArray, list = null) {
+        if (currentIndex < 0) {
+            return list;
+        }
+
+        let arrayElement = inputArray[currentIndex];
+        currentIndex--;
+        return innerArrayToList(inputArray, { value: arrayElement, rest: list});
+    }
+
+    return innerArrayToList(inputArray);
+}
+
+/**
+ * Returns list from array
+ * @param {*} inputArray 
+ * @returns 
+ */
+function arrayToListIterative(inputArray) {
+    if (!Array.isArray(inputArray)) { 
+        throw "input must be array";
+    }
+
+    let listAcc = null;
+
+    for (let i = inputArray.length - 1; i >= 0; i--) {
+        let element = inputArray[i];
+        listAcc = {
+            value: element,
+            rest:listAcc
+        };
+    }
+
+    return listAcc;
+}
+
+/**
+ * Returns array from list
+ * @param {*} inputList 
+ * @returns 
+ */
+function listToArray(inputList) {
+    if (!isList(inputList)) {
+        throw "input must be list data structure";
+    }
+    
+    let arrayAccumulator = [];
+
+    function innerListToArray(list) {
+        arrayAccumulator.push(list.value);
+        if (list.rest) {
```

Comment:
It's generally easier to reason about the base conditions of recursive functions if you put the base conditions in the early returning clauses. For example, make this `if (list.rest === null) {` and move the other logic around to accommodate that.